File: frontend/types/domain.ts
export type ThemePreference = "light" | "dark" | "system";
export type Priority = "urgent" | "high" | "medium" | "low";
export type AssignmentType = "essay" | "homework" | "test" | "presentation" | "reading" | "project";
export type CompletionState = "open" | "completed";
export type SubmissionStatus = "not_started" | "in_progress" | "submitted" | "graded";

export interface Course {
  id: string;
  name: string;
  shortName: string;
  color: string;
}

export interface AssignmentAnalysis {
  difficulty: number;
  urgency: number;
  priorityScore: number;
  explanation: string;
  suggestedSteps: string[];
}

export interface StudySession {
  id: string;
  assignmentId: string;
  title: string;
  startAt: string;
  durationMinutes: number;
  status: "planned" | "completed" | "missed" | "rescheduled";
  source: "canvai" | "manual";
}

export interface Assignment {
  id: string;
  courseId: string;
  title: string;
  description: string;
  type: AssignmentType;
  dueAt: string;
  points: number;
  gradeWeight?: number;
  estimatedMinutes: number;
  actualMinutes?: number;
  priority: Priority;
  submissionStatus: SubmissionStatus;
  missing: boolean;
  completionState: CompletionState;
  scheduledSessionIds: string[];
  analysis: AssignmentAnalysis;
  canvasUrl: string;
}

export interface RoutineBlock {
  id: string;
  day: string;
  activity:
    | "School"
    | "Commute"
    | "Rowing"
    | "Lifting"
    | "Dinner"
    | "Sleep"
    | "Family"
    | "Appointment"
    | "Free time"
    | "Custom";
  startTime: string;
  endTime: string;
  color: string;
}

export interface ConnectionStatus {
  status: "demo" | "connected" | "not_connected" | "error";
  lastSync?: string;
  permissions: string[];
}

export interface CalendarConnection extends ConnectionStatus {
  provider: "google-calendar";
}

export interface CanvasConnection extends ConnectionStatus {
  provider: "canvas";
}

export interface CanvaiInsight {
  id: string;
  title: string;
  body: string;
  impact: string;
  tone: "accent" | "warning" | "success";
}

export interface ScheduleChange {
  id: string;
  kind: "move" | "add" | "remove" | "protect";
  label: string;
  before?: string;
  after?: string;
}

export interface ScheduleProposal {
  id: string;
  command: string;
  summary: string;
  reasoning: string;
  changes: ScheduleChange[];
  status: "preview" | "applied" | "dismissed";
}

export interface CanvaiRecommendation extends CanvaiInsight {
  command: string;
}

export interface Notification {
  id: string;
  title: string;
  body: string;
  timeLabel: string;
  kind: "new" | "changed" | "moved" | "missing" | "resolved" | "connection";
  read: boolean;
}

export interface StudyPreferences {
  earliestTime: string;
  latestTime: string;
  preferredSessionMinutes: number;
  maxSessionMinutes: number;
  breakMinutes: number;
  weekdayWorkloadMinutes: number;
  weekendWorkloadMinutes: number;
  minimumFreeMinutes: number;
  emergencyBufferMinutes: number;
  studyBeforeSchool: boolean;
  studyAtLunch: boolean;
  studyAfterRowing: boolean;
  fridayNight: "avoid" | "light" | "normal";
  saturdayNight: "avoid" | "light" | "normal";
  sundayNight: "avoid" | "light" | "normal";
}

export interface SleepPreferences {
  currentBedtime: string;
  targetBedtime: string;
  wakeTime: string;
  gradualAdjustment: boolean;
  protectSleep: boolean;
  preventLateStudy: boolean;
  athleteRecoveryMode: boolean;
  avoidIntenseAfterTraining: boolean;
}

export interface SubjectPreference {
  courseId: string;
  difficulty: number;
  confidence: number;
  importance: number;
  typicalMinutes: number;
  testDifficulty: number;
  extraTimeMultiplier: number;
}

export interface CalendarPreferences {
  studyCalendar: string;
  busyCalendars: string[];
  automaticPublishing: boolean;
  previewBeforePublishing: boolean;
  notifications: boolean;
  canvaiMayMoveOwnEvents: boolean;
  preserveManualEdits: boolean;
}

export interface AISettings {
  enabled: boolean;
  provider: string;
  analysisDepth: "concise" | "balanced" | "detailed";
  showReasoning: boolean;
  automaticDuration: boolean;
  automaticClassification: boolean;
  useCompletionHistory: boolean;
  askBeforeMajorChanges: boolean;
  estimateFeedback: boolean;
}

export interface UserProfile {
  id: string;
  displayName: string;
  timeZone: string;
  schoolYear: string;
  weekStart: "monday" | "sunday";
  theme: ThemePreference;
}

export interface AssignmentRules {
  testsExtraPriority: boolean;
  startEssaysEarly: boolean;
  splitLongAssignments: boolean;
  createProjectMilestones: boolean;
  missingOverrides: boolean;
  delayLowValueWhenOverloaded: boolean;
  allowSameCourseSessions: boolean;
  includePrepInstructions: boolean;
  rebuildMissedSchedules: boolean;
  keepSundayLight: boolean;
  protectDailyFreeTime: boolean;
}

export interface AppSettings {
  profile: UserProfile;
  study: StudyPreferences;
  sleep: SleepPreferences;
  subjects: SubjectPreference[];
  calendar: CalendarPreferences;
  ai: AISettings;
  rules: AssignmentRules;
}

export interface WeeklyWorkload {
  date: string;
  day: string;
  plannedMinutes: number;
  capacityMinutes: number;
  deadlinePressure: number;
  tests: number;
  writingHeavy: boolean;
}

export interface InsightMetric {
  id: string;
  label: string;
  value: string;
  change: string;
  explanation: string;
  adjustment: string;
}

export interface DemoBootstrap {
  referenceDate: string;
  courses: Course[];
  assignments: Assignment[];
  sessions: StudySession[];
  routine: RoutineBlock[];
  notifications: Notification[];
  workload: WeeklyWorkload[];
  settings: AppSettings;
  canvasConnection: CanvasConnection;
  calendarConnection: CalendarConnection;
}

File: backend/app/schemas/domain.py
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field, field_validator

from app.schemas.phase2 import validate_timezone


class CourseSchema(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: str
    name: str
    short_name: str
    color: str


class AssignmentAnalysis(BaseModel):
    difficulty: int = Field(ge=1, le=5)
    urgency: int = Field(ge=1, le=5)
    priority_score: int = Field(ge=0, le=100)
    explanation: str
    suggested_steps: list[str]


class AssignmentSchema(BaseModel):
    id: str
    course_id: str
    title: str
    description: str
    type: str
    due_at: datetime | None
    points: float
    grade_weight: float | None = None
    estimated_minutes: int
    actual_minutes: int | None = None
    priority: str
    submission_status: str
    missing: bool
    completion_state: str
    scheduled_session_ids: list[str]
    analysis: AssignmentAnalysis
    canvas_url: str


class AssignmentUpdate(BaseModel):
    title: str | None = Field(default=None, min_length=3, max_length=240)
    estimated_minutes: int | None = Field(default=None, ge=10, le=600)
    completion_state: Literal["open", "completed"] | None = None
    submission_status: Literal["not_started", "in_progress", "submitted", "graded"] | None = None
    missing: bool | None = None


class StudySessionSchema(BaseModel):
    id: str
    assignment_id: str
    title: str
    start_at: datetime
    duration_minutes: int
    status: str
    source: str


class RoutineBlockSchema(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: str
    day: str
    activity: str
    start_time: str
    end_time: str
    color: str


class NotificationSchema(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: str
    title: str
    body: str
    time_label: str
    kind: str
    read: bool


class UserProfileSchema(BaseModel):
    id: str
    display_name: str = Field(min_length=1, max_length=120)
    time_zone: str
    school_year: str
    week_start: Literal["monday", "sunday"]
    theme: Literal["light", "dark", "system"]

    _timezone_is_valid = field_validator("time_zone")(validate_timezone)


class StudyPreferencesSchema(BaseModel):
    earliest_time: str
    latest_time: str
    preferred_session_minutes: int = Field(ge=10, le=180)
    max_session_minutes: int = Field(ge=10, le=240)
    break_minutes: int = Field(ge=0, le=60)
    weekday_workload_minutes: int = Field(ge=0, le=600)
    weekend_workload_minutes: int = Field(ge=0, le=600)
    minimum_free_minutes: int = Field(ge=0, le=600)
    emergency_buffer_minutes: int = Field(ge=0, le=180)
    study_before_school: bool
    study_at_lunch: bool
    study_after_rowing: bool
    friday_night: str
    saturday_night: str
    sunday_night: str


class SleepPreferencesSchema(BaseModel):
    current_bedtime: str
    target_bedtime: str
    wake_time: str
    gradual_adjustment: bool
    protect_sleep: bool
    prevent_late_study: bool
    athlete_recovery_mode: bool
    avoid_intense_after_training: bool


class SubjectPreferenceSchema(BaseModel):
    course_id: str
    difficulty: int = Field(ge=1, le=5)
    confidence: int = Field(ge=1, le=5)
    importance: int = Field(ge=1, le=5)
    typical_minutes: int = Field(ge=5, le=600)
    test_difficulty: int = Field(ge=1, le=5)
    extra_time_multiplier: float = Field(ge=1, le=3)


class CalendarPreferencesSchema(BaseModel):
    study_calendar: str
    busy_calendars: list[str]
    automatic_publishing: bool
    preview_before_publishing: bool
    notifications: bool
    canvai_may_move_own_events: bool
    preserve_manual_edits: bool


class AISettingsSchema(BaseModel):
    enabled: bool
    provider: str
    analysis_depth: Literal["concise", "balanced", "detailed"]
    show_reasoning: bool
    automatic_duration: bool
    automatic_classification: bool
    use_completion_history: bool
    ask_before_major_changes: bool
    estimate_feedback: bool


class AssignmentRulesSchema(BaseModel):
    tests_extra_priority: bool
    start_essays_early: bool
    split_long_assignments: bool
    create_project_milestones: bool
    missing_overrides: bool
    delay_low_value_when_overloaded: bool
    allow_same_course_sessions: bool
    include_prep_instructions: bool
    rebuild_missed_schedules: bool
    keep_sunday_light: bool
    protect_daily_free_time: bool


class AppSettingsSchema(BaseModel):
    profile: UserProfileSchema
    study: StudyPreferencesSchema
    sleep: SleepPreferencesSchema
    subjects: list[SubjectPreferenceSchema]
    calendar: CalendarPreferencesSchema
    ai: AISettingsSchema
    rules: AssignmentRulesSchema


class WorkloadSchema(BaseModel):
    date: datetime
    day: str
    planned_minutes: int
    capacity_minutes: int
    deadline_pressure: int
    tests: int
    writing_heavy: bool


class InsightMetricSchema(BaseModel):
    id: str
    label: str
    value: str
    change: str
    explanation: str
    adjustment: str


class ConnectionSchema(BaseModel):
    provider: str
    status: str
    last_sync: str | None = None
    permissions: list[str]


class WorkspaceBootstrapSchema(BaseModel):
    courses: list[CourseSchema]
    assignments: list[AssignmentSchema]
    sessions: list[StudySessionSchema]
    routine: list[RoutineBlockSchema]
    notifications: list[NotificationSchema]
    workload: list[WorkloadSchema]
    settings: AppSettingsSchema


class CanvaiProposalRequest(BaseModel):
    command: str = Field(min_length=3, max_length=180)


class ScheduleChangeSchema(BaseModel):
    id: str
    kind: Literal["move", "add", "remove", "protect"]
    label: str
    before: str | None = None
    after: str | None = None


class ScheduleProposalSchema(BaseModel):
    id: str
    command: str
    summary: str
    reasoning: str
    changes: list[ScheduleChangeSchema]
    status: Literal["preview", "applied", "dismissed"] = "preview"


class StatusSchema(BaseModel):
    status: str
    service: str
    database: str | None = None
